### PR TITLE
SWC-5724, SWC-5684, SWC-5762

### PR DESCRIPTION
**Pull request by nickgros**
[SWC-5724](https://sagebionetworks.jira.com/browse/SWC-5724)

When a schema author defines an enumeration using `anyOf`, the client-side validator doesn't recognize that it's an enumeration, so the error message when making an invalid selection is verbose and hard to understand. This change tries to recognize when we see errors that represent this situation, and transforms the errors to simplify what we show the user.

Before:
image

After:
image

[SWC-5684](https://sagebionetworks.jira.com/browse/SWC-5684) incorporates a collapsible description element from the Figma design:

https://user-images.githubusercontent.com/17580037/131906800-5c8bd4c8-6100-448b-97a7-849144c42198.mov

[SWC-5762](https://sagebionetworks.jira.com/browse/SWC-5762) initializes arrays as empty, rather than null, which allows users to immediately interact with the form, rather than having to click 'Add Value' for each array (in some complex schemas, this is a lot of fields). We make this work by stripping null-ish array values and then removing empty arrays when we submit a new set of annotations.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,7 +25,7 @@
     "@material-ui/icons": "^4.11.2",
     "@material-ui/lab": "^4.0.0-alpha.57",
     "@rjsf/core": "^3.0.0",
-    "@sage-bionetworks/rjsf-core": "^3.1.3",
+    "@sage-bionetworks/rjsf-core": "^3.1.4",
```
Comment (nickgros) [inline package.json:28]: Fixes type issues that were in RJSF

File: src/__tests__/lib/containers/entity/annotations/SchemaDrivenAnnotationsEditor.test.tsx [hunk 2]
```diff
@@ -501,4 +525,60 @@ describe('SchemaDrivenAnnotationEditor tests', () => {
...
+
+    // Now hide the description table
+    userEvent.click(moreInfoButton)
+
+    await waitFor(() =>
+      expect(
+        screen.queryByText('Test description for country property'),
+      ).not.toBeVisible(),
+    )
+  })
```
Comment (nickgros) [inline src/__tests__/lib/containers/entity/annotations/SchemaDrivenAnnotationsEditor.test.tsx:554]: Test for SWC-5684

File: src/__tests__/lib/containers/entity/annotations/SchemaDrivenAnnotationsEditor.test.tsx [hunk 3]
```diff
@@ -501,4 +525,60 @@ describe('SchemaDrivenAnnotationEditor tests', () => {
       { exact: false },
     )
   })
+
+  it('Shows a schema description and type when help is clicked', async () => {
```
Comment (nickgros) [inline src/__tests__/lib/containers/entity/annotations/SchemaDrivenAnnotationsEditor.test.tsx:529]: Test for SWC-5684

File: src/__tests__/lib/containers/entity/annotations/SchemaDrivenAnnotationsEditor.test.tsx [hunk 4]
```diff
@@ -501,4 +525,60 @@ describe('SchemaDrivenAnnotationEditor tests', () => {
...
+    userEvent.click(moreInfoButton)
+
+    await waitFor(() =>
+      expect(
+        screen.queryByText('Test description for country property'),
+      ).not.toBeVisible(),
+    )
+  })
+
+  it('Initializes an empty array but does not submit null data', async () => {
```
Comment (nickgros) [inline src/__tests__/lib/containers/entity/annotations/SchemaDrivenAnnotationsEditor.test.tsx:556]: Test for SWC-5762

File: src/lib/containers/entity/annotations/AnnotationEditorUtils.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
+import { AjvError } from '@sage-bionetworks/rjsf-core'
```
Comment (nickgros) [inline src/lib/containers/entity/annotations/AnnotationEditorUtils.ts:1]: The functions in this file were pulled out of the SchemaDrivenAnnotationEditor component file. They are also now tested separately.

File: src/lib/containers/entity/annotations/AnnotationEditorUtils.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+  // Fixing anyOf errors
+  // Group the errors by the property that the error applies to
+  const grouped = groupBy(errors, error => error.property)
+  Object.keys(grouped).map(property => {
+    const errorGroup = grouped[property]
+
+    // First, see if it is an anyOf error
+    const hasAnyOfError = errorGroup.some(
+      e => e.message === 'should match some schema in anyOf',
```
Comment (jay-hodgson) [inline src/lib/containers/entity/annotations/AnnotationEditorUtils.ts:58]: Hope this error message does not change!

File: src/lib/containers/entity/annotations/CustomArrayFieldTemplate.tsx [hunk 7]
```diff
@@ -1,40 +1,62 @@
...
  * Used just to apply custom styling.
  */
 export function CustomArrayFieldTemplate<T>(props: ArrayFieldTemplateProps<T>) {
   const { DescriptionField } = props
+
+  const isAdditionalProperty = props.schema[utils.ADDITIONAL_PROPERTY_FLAG]
+
+  useEffect(() => {
+    if (props.items.length === 0) {
+      props.onAddClick()
```
Comment (jay-hodgson) [inline src/lib/containers/entity/annotations/CustomArrayFieldTemplate.tsx:18]: Removing pain points

**Review by jay-hodgson — APPROVED** [review]

**Pull request merged**